### ISSUE TITLE
ConfuserEx 0.5.0 seems to be deobfuscator

### ISSUE THREAD

**Issue by vietnguyen09** [yck1509/ConfuserEx#305]
Hi developer,

You might need look over this page.

http://hackandmodz.net/threads/crackme-unpackme-simple-confuserex-0-5-0.8617/

With me, it's a very sad news. 
How can I chage the encrytion method by myself?

I'm looking forward to waiting new version.

**Comment by vietnguyen09** [comment]
How can I chage the encrytion method by myself?

**Comment by picrap** [comment]
You probably should read that thread more carefully, because method names are still obfuscated. And anyway, an obfuscator is always reversible in some way, just because it has to be run by the CLR :wink:

**Comment by atom0s** [comment]
ConfuserEx is a public and open source obfuscator. No matter what changes are ever made to it, there will always be easy methods to debofuscate anything used by it. Since crackers and such can see what the application does in the actual source, there is pretty much no protection gained using ConfuserEx except to weed out the newbie crackers that cannot do anything without automated tools.

**Comment by m5112638** [comment]
I can see renaming is pretty useful, which makes it much harder to understand the code, giving the renaming process is not reversible.

**Comment by picrap** [comment]
Note to @atom0s : all well-known cryptographic algorithm are open source, and still, they're used and reliable. A lot of applications are open source and secure anyway. This is even better, because everyone can check their quality. What is important here is the "noise" added by ConfuserEx: noise in code flow, noise in names, and since all of this is random, it can not be reversed. So yes it works.

**Comment by atom0s** [comment]
I never said it doesn't work, just that it is not something people should use as a reliable method to protect sensitive data / source code. I personally use ConfuserEx for some projects because it helps keep out the newbie crackers and such. However, if I ever created something that was very sensitive or something I did not want to have source code shared/seen, ConfuserEx is not the tool I would use for it.

Projects like ConfuserEx help weed out the drag-and-drop tool users that want to steal your work or crack a paid application but have no idea how to do any reversing themselves. People with more experience however do not have much trouble undoing what ConfuserEx (and similar projects) does. There are automated tools to defeat every part of ConfuserEx already, regardless of what changes are made to it currently. 

However, the one advantage to being open source is that everyone has the ability to alter it to their own needs and change some protections or add more all while not re-sharing the source. Sadly though, that is taken advantage of, as seen with various rip-offs of this project that are solely out for profit off others work.

tldr; ConfuserEx works great to keep wannabe crackers out of your source code, but does little to really protect your work from being cracked/deobfuscated.

**Comment by MurariAlex** [comment]
Take [de4dot deobfuscator](https://github.com/0xd4d/de4dot) for example, it supports the majority of closed source obfuscators, but doesn't support ConfuserEx.
If you remove the watermarking of ConfuserEx (or change it's signature), it will be real hard for a wannabe cracker to find out which protection you are using, and gather all the tools necessary to undo it.
However, a skilled cracker will break into your code and find out the deepest secrets of your software, the question is: How much people want your code ckracked?
C#, VB.NET, and other CLR dependent languages has too many metadata in the IL.
If you want protection, use C++, it's way more secure, but, then again, how much people want your code cracked?

**Comment by romeop555** [comment]
Interesting, how can you remove the watermarking of ConfuserEx (or change it's signature)?

**Comment by MurariAlex** [comment]
In this [line of code](https://github.com/yck1509/ConfuserEx/blob/master/Confuser.Core/ConfuserEngine.cs#L306) in ConfuserEngine, change "ConfusedByAttribute" to whatever you like, or remove the relevant parts to avoid watermarking.

**Comment by vietnguyen09** [comment]
After few weeks I went around and did some researched and I can saw ConfuserEx was a very strong obfuscator tool, first of all we need to consider about deobfuscator tool for ConfuserEx is none, secondly, it strong because it's opensource and we can custome source to make your own obfuscator, when we do this, we can protect our source from scrip kids who are always use deobfuscator tools to steal source code. Finally, really thanks ConfuserEx team for everything they have done for us.

**Comment by atom0s** [comment]
There are multiple tools to deobfuscate ConfuserEx.

**Comment by vietnguyen09** [comment]
@atom0s  but when you change secret key in ConfuserEx source code and remove watermark, these tools will fail.

**Comment by SandPox** [comment]
@vietnguyen09 and you don't think they'll rename it back when they want to deobfuscator it?

**Comment by atom0s** [comment]
@vietnguyen09 at that point if someone still needs it cracked they can adjust the tools as needed to find your alterations and still work or they can just use an IL editor and alter the changes back to the default ConfuserEx strings. It's not hard to redo them.

**Comment by m5112638** [comment]
I have to admit I don't know how ConfuserEx does renaming. If it renames each name to something random, not based on some algorithm (which may be reversible), then all name information is lost and there is no way to rename the random names back to its original names. This is a powerful protection.

Again I am not sure how other protections work. They may be able to be reversed because no matter how complicated these protections are they still need to do what the original code is intended to do.

**Comment by J2TeaM** [comment]
@m5112638 Yes. Confuser is using generate random name method as other obfuscators.

![image](https://cloud.githubusercontent.com/assets/5250117/9565147/59dd4722-4ee8-11e5-9eb5-7ca53eb5edb4.png)

They can change the names of methods, variables to an understandable string but cannot get back the original name.

**Comment by m5112638** [comment]
Thanks @J2TeaM.

@vietnguyen09 how do you change the secret key?

**Comment by vietnguyen09** [comment]
@m5112638 You can find module name 'koi' and change this, that all I know after few days I've been looking the way for mod ConfuserEx

**Comment by SandPox** [comment]
you can open the protected file with ILSpy to see the protector function name... then open the file with hex editor then replace those function name... as far as file still can run then you're do it right

**Comment by yck1509** [comment]
Hi,

As with every other obfuscators, a determined cracker can always crack the application given enough time. What ConfuserEx does is to increase the time cost for them, and the difficulty of cracking through protections of ConfuserEx was one of the highest of publicly available obfuscator.

It is known that there are tools to deobfuscate ConfuserEx, however, many tools cannot handle advanced parameters of ConfuserEx, and only a small amount of crackers know how to modify a tool to deal with modifications to ConfuserEx. I cannot assure the absolute security, but I suppose the provided protections are quite good.

**Comment by DeathGodAoTTG** [comment]
@atom0s what are the names of these tools?
